lib/puppet/transaction/persistence.rb

```diff
@@ -87,7 +87,16 @@ def load
 
   # Save data from internal class to persistence store on disk.
   def save
-    Puppet::Util::Yaml.dump(@new_data, Puppet[:transactionstorefile])
+    converted_data = Puppet::Pops::Serialization::ToDataConverter.convert(
+      @new_data, {
+        symbol_as_string: false,
+        local_reference: false,
+        type_by_reference: true,
+        force_symbol: true,
+        message_prefix: to_s
+      }
+    )
+    Puppet::Util::Yaml.dump(converted_data, Puppet[:transactionstorefile])
```

Comment:
This still raises a warning for me when converting objects:

```
Warning: #<Puppet::Transaction::Persistence:0x0000000005250440>['resources']['Notify[deferred]']['parameters']['message']['system_value'][0] contains a Puppet::Pops::Types::PBinaryType::Binary value. It will be converted to the String 'aGVsbG8K'
```

maybe we can guard against that as well?

Reply:
Something like this should work
```diff
diff --git a/lib/puppet/pops/serialization/to_data_converter.rb b/lib/puppet/pops/serialization/to_data_converter.rb
index 462f612774..ea3c82e491 100644
--- a/lib/puppet/pops/serialization/to_data_converter.rb
+++ b/lib/puppet/pops/serialization/to_data_converter.rb
@@ -117,7 +117,11 @@ module Serialization
           { PCORE_TYPE_KEY => PCORE_TYPE_SENSITIVE, PCORE_VALUE_KEY => to_data(value.unwrap) }
         end
       else
-        unknown_to_data(value)
+        if @rich_data
+          value_to_data_hash(value)
+        else
+          @option_that_silences_warnings ? unknown_to_string(value) : unknown_to_string_with_warning(value)
+        end
       end
     end
 
```